pep-9999.rst

```diff
@@ -0,0 +1,89 @@
+PEP: 9999
+Title: Recording provenance of installed packages
+Author: Fridolin Pokorny <fridolin.pokorny at datadoghq.com>, Trishank Karthik Kuppusamy <trishank.kuppusamy at datadoghq.com>, Donald Stufft <donald.stufft at datadoghq.com>, Ofek Lev <ofek.lev at ofek.lev@datadoghq.com datadoghq.com>
+Sponsor: Paul Moore <p.f.moore at gmail.com>
+PEP-Delegate:
+Discussions-To: TODO
+Status: Draft
+Type: Process
+Content-Type: text/x-rst
+Created: 30-Jan-2002
+Post-History:
+
+
+Abstract
+========
+
+This PEP describes a way to record provenance of Python packages installed.
+This record is created by an installer and is available to users in a form of a
+JSON file ``direct_url.json`` in ``.dist-info`` directory. This PEP extends
+`PEP-610 <https://peps.python.org/pep-0610/>`_.
+
+
+Motivation
+==========
+
+Installing a Python package involves downloading the package from a source and
+extracting its content to an appropriate place. After the installation process
+is done, information about the artifact used as well as its source is generally
+lost. Nevertheless, there are use cases for keeping records of artifacts used
+for installing packages and their provenance.
+
+Python wheels can be built with different compiler flags or supporting
+different wheel tags.  In both cases, users might get into a situation in which
+multiple wheels might be considered by installers (possibly from different
+package indexes) and immediately finding out which artifact was actually used
+during the installation might be helpful. This way, tools reporting software
+installed, such as tools reporting SBOM (Software Bill of Material), might give
+more accurate reports.
+
+The motivation described in this PEP is a direct extension to `PEP-610
+<https://peps.python.org/pep-0610/>`_. Besides stating information about
+packages installed from an URL, installers SHOULD record information also
+for packages installed from Python package indexes.
+
+
+Examples
+========
+
+An example of a ``direct_url.json`` file in the ``.dist-info`` directory
+alongside files stated in `PEP-376 <https://peps.python.org/pep-0376/>`_ and
+further adjusted in `PEP-627 <https://peps.python.org/pep-0627/>`_. The
+specified file stores a JSON describing artifact used to install a package as
+well as its source.  The record is similar to ``direct_url.json`` described in
+`PEP-610 <https://peps.python.org/pep-0610/>`_:
+
+.. code-block: json
+
+  {
+    "archive_info": {
+      "hash": "sha256=714ac14496c3e68c99c29b00845f7a2b85f3bb6f1078fd9f72fd20f0570002b2"
+    },
+    "url": "https://files.pythonhosted.org/packages/ed/35/a31aed2993e398f6b09a790a181a7927eb14610ee8bbf02dc14d31677f1c/packaging-23.0-py3-none-any.whl"
+  }
+
+The JSON document is stating the following entries:
+
+* ``archive_info.hash`` MUST be present with a value of ``<hash-algorithm>=<expected-hash>``,
+  currently supported hash algorithm is only ``sha256``.
```

Comment:
Sounds reasonable, I'm not sure how the ``archive_info.hash`` should look like in such a case as ``<hash-algorithm>`` is directly part of its value (maybe comma separated?). This might also open discussion how it should look like in already existing support for ``direct_url.json``.